src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -454,6 +462,46 @@ protected override int AddVariable(VariableIdentifier identifier)
             return _variables.Add(identifier);
         }
 
+        // Note: this is for placeholders to unvisited expressions
+        private void AddPlaceholder(BoundValuePlaceholderBase placeholder, BoundExpression unvisited)
+        {
+            _placeholdersToUnvisitedExpressionOpt ??= PooledDictionary<BoundValuePlaceholderBase, BoundExpression>.GetInstance();
+            _placeholdersToUnvisitedExpressionOpt.Add(placeholder, unvisited);
+        }
+
+        // Note: this is for placeholders to unvisited expressions
+        private void RemovePlaceholder(BoundValuePlaceholderBase placeholder)
+        {
+            Debug.Assert(_placeholdersToUnvisitedExpressionOpt is not null);
+            _placeholdersToUnvisitedExpressionOpt.Remove(placeholder);
+        }
+
+        // Note: this is for placeholders to unvisited expressions
+        private bool TryReplacePlaceholder(BoundValuePlaceholderBase placeholder, [NotNullWhen(true)] out BoundExpression? unvisited)
+        {
+            if (_placeholdersToUnvisitedExpressionOpt is not null &&
+                _placeholdersToUnvisitedExpressionOpt.TryGetValue(placeholder, out unvisited))
+            {
+                return true;
```

Comment:
We should be able to do `Debug.Assert(_analyzedNullabilityMapOpt?.ContainsKey(unvisited) != true);` here.

Reply:
I hit a problem with this in simple tests like `System.Console.Write((x, new C()[^1]));` because tuples are visited twice (we visit the source tuple once, then the converted tuple, see `VisitConvertedTupleLiteral`).
We probably would hit a problem in a loop as well (when nullable walker needs to make multiple passes through the code to get a stable answer).
So "unvisited" here doesn't mean "never visited" but rather "whoever substitutes this placeholder is responsible for visiting".

Reply:
Ah yes, tuples, the bane of my existence in nullability.